// uniwind@1.9.0/dist/common/core/native/store.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UniwindStore = void 0;
var _reactNative = require("react-native");
var _consts = require("../../common/consts");
var _listener = require("../listener");
var _parsers = require("./parsers");
var _runtime = require("./runtime");
const emptyState = {
  styles: {},
  dependencies: [],
  dependencySum: 0
};
class UniwindStoreBuilder {
  runtime = _runtime.UniwindRuntime;
  vars = {};
  stylesheet = {};
  cache = {};
  getStyles(className, componentProps, state, uniwindContext) {
    if (className === void 0 || className === "") {
      return emptyState;
    }
    const isScopedTheme = uniwindContext.scopedTheme !== null;
    const cacheKey = `${className}${state?.isDisabled ?? false}${state?.isFocused ?? false}${state?.isPressed ?? false}${isScopedTheme}${uniwindContext.rtl ?? ""}`;
    const cache = this.cache[uniwindContext.scopedTheme ?? this.runtime.currentThemeName];
    if (!cache) {
      return emptyState;
    }
    if (cache.has(cacheKey)) {
      return cache.get(cacheKey);
    }
    const result = this.resolveStyles(className, componentProps, state, uniwindContext);
    if (!result.hasDataAttributes) {
      cache.set(cacheKey, result);
      _listener.UniwindListener.subscribe(() => cache.delete(cacheKey), result.dependencies, {
        once: true
      });
    }
    return result;
  }
  reinit = (generateStyleSheetCallback, themes) => {
    const config = generateStyleSheetCallback(this.runtime);
    const {
      scopedVars,
      stylesheet,
      vars
    } = config;
    const platform = this.getCurrentPlatform();
    const commonPlatform = platform.includes("tv") ? _consts.Platform.TV : _consts.Platform.Native;
    const commonPlatformVars = scopedVars[`${_consts.UNIWIND_PLATFORM_VARIABLES}${commonPlatform}`];
    const platformVars = scopedVars[`${_consts.UNIWIND_PLATFORM_VARIABLES}${platform}`];
    if (commonPlatformVars) {
      Object.assign(vars, commonPlatformVars);
    }
    if (platformVars) {
      Object.assign(vars, platformVars);
    }
    this.stylesheet = stylesheet;
    this.vars = Object.fromEntries(themes.map(theme => {
      const clonedVars = Object.create(vars);
      const themeVars = scopedVars[`${_consts.UNIWIND_THEME_VARIABLES}${theme}`];
      if (themeVars) {
        Object.assign(clonedVars, themeVars);
      }
      return [theme, clonedVars];
    }));
    this.cache = Object.fromEntries(themes.map(theme => [theme, /* @__PURE__ */new Map()]));
    if (__DEV__) {
      _listener.UniwindListener.notifyAll();
    }
  };
  resolveStyles(classNames, componentProps, state, uniwindContext) {
    const resultGetters = {};
    const theme = uniwindContext.scopedTheme ?? this.runtime.currentThemeName;
    let vars = this.vars[theme];
    const originalVars = vars;
    let hasDataAttributes = false;
    const dependencies = /* @__PURE__ */new Set();
    let dependencySum = 0;
    const bestBreakpoints = /* @__PURE__ */new Map();
    const isScopedTheme = uniwindContext.scopedTheme !== null;
    for (const className of classNames.split(" ")) {
      if (!(className in this.stylesheet)) {
        continue;
      }
      for (const style of this.stylesheet[className]) {
        if (style.dependencies) {
          style.dependencies.forEach(dep => {
            if (dep === _consts.StyleDependency.Theme && isScopedTheme) {
              return;
            }
            dependencies.add(dep);
            dependencySum |= 1 << dep;
          });
        }
        if (style.dataAttributes !== null) {
          hasDataAttributes = true;
        }
        if (style.minWidth > this.runtime.screen.width || style.maxWidth < this.runtime.screen.width || style.theme !== null && theme !== style.theme || style.orientation !== null && this.runtime.orientation !== style.orientation || style.rtl !== null && !this.validateDir(style.rtl, uniwindContext) || style.active !== null && state?.isPressed !== style.active || style.focus !== null && state?.isFocused !== style.focus || style.disabled !== null && state?.isDisabled !== style.disabled || style.dataAttributes !== null && !this.validateDataAttributes(style.dataAttributes, componentProps)) {
          continue;
        }
        for (const [property, valueGetter] of style.entries) {
          const previousBest = bestBreakpoints.get(property);
          if (previousBest && (previousBest.minWidth > style.minWidth || previousBest.complexity > style.complexity || previousBest.importantProperties.includes(property))) {
            continue;
          }
          if (property[0] === "-") {
            if (vars === originalVars) {
              vars = Object.create(originalVars);
            }
            vars[property] = valueGetter;
          } else {
            resultGetters[property] = valueGetter;
          }
          bestBreakpoints.set(property, style);
        }
      }
    }
    const result = Object.fromEntries(Object.entries(resultGetters).map(([property, valueGetter]) => [property, valueGetter(vars)]));
    if (result.lineHeight !== void 0 && result.lineHeight < 6) {
      result.lineHeight *= result.fontSize;
    }
    if (result.boxShadow !== void 0) {
      result.boxShadow = (0, _parsers.parseBoxShadow)(result.boxShadow);
    }
    if (result.visibility === "hidden") {
      result.display = "none";
    }
    if (result.borderStyle !== void 0 && result.borderColor === void 0) {
      result.borderColor = "#000000";
    }
    if (result.outlineStyle !== void 0 && result.outlineColor === void 0) {
      result.outlineColor = "#000000";
    }
    if (result.fontVariant !== void 0) {
      result.fontVariant = (0, _parsers.parseFontVariant)(result.fontVariant);
    }
    (0, _parsers.parseTransformsMutation)(result);
    if (result.experimental_backgroundImage !== void 0) {
      result.experimental_backgroundImage = (0, _parsers.resolveGradient)(result.experimental_backgroundImage);
    }
    if (result.textShadow !== void 0) {
      (0, _parsers.parseTextShadowMutation)(result);
    }
    return {
      styles: result,
      dependencies: Array.from(dependencies),
      dependencySum,
      hasDataAttributes
    };
  }
  validateDataAttributes(dataAttributes, props = {}) {
    for (const [attribute, expectedAttributeValue] of Object.entries(dataAttributes)) {
      const attributeValue = props[attribute];
      if (expectedAttributeValue === "true") {
        if (attributeValue !== true && attributeValue !== "true") {
          return false;
        }
        continue;
      }
      if (expectedAttributeValue === "false") {
        if (attributeValue !== false && attributeValue !== "false") {
          return false;
        }
        continue;
      }
      if (attributeValue !== expectedAttributeValue) {
        return false;
      }
    }
    return true;
  }
  validateDir(rtl, uniwindContext) {
    if (uniwindContext.rtl !== null) {
      return rtl === uniwindContext.rtl;
    }
    return rtl === this.runtime.rtl;
  }
  getCurrentPlatform() {
    const platform = _reactNative.Platform.OS;
    if (platform === "android") {
      return _reactNative.Platform.isTV ? _consts.Platform.AndroidTV : _consts.Platform.Android;
    }
    if (platform === "ios") {
      return _reactNative.Platform.isTV ? _consts.Platform.AppleTV : _consts.Platform.iOS;
    }
    return platform;
  }
}
const UniwindStore = exports.UniwindStore = new UniwindStoreBuilder();
_reactNative.Dimensions.addEventListener("change", ({
  window
}) => {
  const newOrientation = window.width > window.height ? _consts.Orientation.Landscape : _consts.Orientation.Portrait;
  const orientationChanged = UniwindStore.runtime.orientation !== newOrientation;
  UniwindStore.runtime.screen = {
    width: window.width,
    height: window.height
  };
  UniwindStore.runtime.orientation = newOrientation;
  _listener.UniwindListener.notify([...(orientationChanged ? [_consts.StyleDependency.Orientation] : []), _consts.StyleDependency.Dimensions]);
});
