// uniwind@1.9.0/dist/common/bundler/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UniwindBundlerConfig = void 0;
var _css = require("@/bundler/artifacts/css");
var _dts = require("@/bundler/artifacts/dts");
var _cssVisitor = require("@/bundler/css-visitor");
var _consts = require("@/common/consts");
var _utils = require("@/common/utils");
var _path = _interopRequireDefault(require("path"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
class UniwindBundlerConfig {
  constructor(config, platform) {
    this.config = config;
    this.platform = platform;
  }
  static fromMetroConfig(config, platform) {
    const getPlatform = () => {
      if (!(0, _utils.isDefined)(platform)) {
        return _consts.Platform.Native;
      }
      if (!config.isTV) {
        return platform;
      }
      if (platform === _consts.Platform.Android) {
        return _consts.Platform.AndroidTV;
      }
      if (platform === _consts.Platform.iOS) {
        return _consts.Platform.AppleTV;
      }
      throw new Error(`Platform ${platform} not supported`);
    };
    if (typeof config === "undefined") {
      throw new Error("Uniwind: You need to pass second parameter to withUniwindConfig");
    }
    if (typeof config.cssEntryFile === "undefined") {
      throw new Error('Uniwind: You need to pass css css entry file to withUniwindConfig, e.g. withUniwindConfig(config, { cssEntryFile: "./global.css" })');
    }
    return new UniwindBundlerConfig(config, getPlatform());
  }
  static fromViteConfig(config) {
    return new UniwindBundlerConfig(config, _consts.Platform.Web);
  }
  static fromCliConfig(config) {
    if (typeof config.cssEntryFile === "undefined") {
      throw new Error("Uniwind: You need to pass css entry file, e.g. uniwind generate-artifacts --css ./global.css. Run uniwind generate-artifacts --help for usage.");
    }
    return new UniwindBundlerConfig(config, _consts.Platform.Web);
  }
  get cssPath() {
    return _path.default.join(process.cwd(), this.config.cssEntryFile);
  }
  get themes() {
    return Array.from(/* @__PURE__ */new Set(["light", "dark", ...(this.config.extraThemes ?? [])]));
  }
  get cssVisitor() {
    return new _cssVisitor.UniwindCSSVisitor(this);
  }
  get polyfills() {
    return this.config.polyfills;
  }
  get stringifiedThemes() {
    return `[${this.themes.map(theme => `'${theme}'`).join(", ")}]`;
  }
  toMetroConfig() {
    return this.config;
  }
  async generateArtifacts(cssArtifactPath) {
    await (0, _css.buildCSS)(this.themes, this.config.cssEntryFile, cssArtifactPath);
    (0, _dts.buildDtsFile)(this.config.dtsFile ?? "uniwind-types.d.ts", this.stringifiedThemes);
  }
}
exports.UniwindBundlerConfig = UniwindBundlerConfig;
