// uniwind@1.9.0/dist/common/core/config/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Uniwind = void 0;
var _consts = require("../../common/consts");
var _utils = require("../../common/utils");
var _listener = require("../listener");
var _logger = require("../logger");
var _web = require("../web");
var _config = require("./config.common");
class UniwindConfigBuilder extends _config.UniwindConfigBuilder {
  cssRules;
  constructor() {
    super();
  }
  updateCSSVariables(theme, variables) {
    if (typeof document === "undefined") {
      return;
    }
    const uniwindRules = this.getUniwindDynamicCSSRules();
    Object.entries(variables).forEach(([varName, varValue]) => {
      if (!varName.startsWith("--") && __DEV__) {
        _logger.Logger.error(`CSS variable name must start with "--", instead got: ${varName}`);
      }
      const existingRules = Object.fromEntries(uniwindRules.map(rule => [rule.theme, (0, _web.getWebVariable)(varName, {
        scopedTheme: rule.theme,
        rtl: null
      })]));
      uniwindRules.forEach(rule => {
        if (rule.theme === theme) {
          rule.style.setProperty(varName, typeof varValue === "number" ? `${varValue}px` : varValue);
          return;
        }
        rule.style.setProperty(varName, existingRules[rule.theme] ?? null);
      });
    });
    _listener.UniwindListener.notify([_consts.StyleDependency.Variables]);
  }
  __reinit(generateStyleSheetCallback, themes) {
    const oldThemes = this.themes;
    super.__reinit(generateStyleSheetCallback, themes);
    if ((0, _utils.arrayEquals)(themes, oldThemes)) {
      return;
    }
    this.cssRules = void 0;
    if (typeof document !== "undefined") {
      document.querySelector("#uniwind-dynamic-styles")?.remove();
    }
  }
  getUniwindDynamicCSSRules() {
    if (this.cssRules) {
      return this.cssRules;
    }
    if (typeof document === "undefined") {
      return [];
    }
    const styleElement = document.createElement("style");
    styleElement.innerText = this.themes.reduce((acc, theme) => {
      return `${acc}.${theme}{}`;
    }, "");
    styleElement.setAttribute("id", "uniwind-dynamic-styles");
    document.head.appendChild(styleElement);
    const cssRules = Array.from(styleElement.sheet?.cssRules ?? []).filter(rule => "selectorText" in rule && "style" in rule).map(rule => ({
      style: rule.style,
      theme: rule.selectorText.replace(".", "")
    }));
    this.cssRules = cssRules;
    return cssRules;
  }
}
const Uniwind = exports.Uniwind = new UniwindConfigBuilder();
