// uniwind@1.9.0/dist/vite/index.mjs
import { normalizePath } from '@tailwindcss/node';
import { createRequire } from 'node:module';
import path from 'path';
import { U as UniwindBundlerConfig } from '../shared/uniwind.WHb-3yIn.mjs';
import 'fs';
import 'lightningcss';

const dirname = typeof __dirname !== "undefined" ? __dirname : import.meta.dirname;
const componentPath = path.resolve(
  dirname,
  "../module/components/web/index.js"
);
const styleSheetPath = path.resolve(
  dirname,
  "../module/components/web/createOrderedCSSStyleSheet.js"
);
const cssArtifactPath = path.resolve(dirname, "../../uniwind.css");
const require = createRequire(import.meta.url);
const viteVersion = require("vite/package.json").version;
const isVite8 = Number(viteVersion.split(".")[0]) >= 8;
const resolveOrderedCSSStyleSheet = (source, importer) => {
  const normalizedSource = normalizePath(source);
  const isTarget = source === "./createOrderedCSSStyleSheet" || normalizedSource.endsWith("react-native-web/dist/exports/StyleSheet/dom/createOrderedCSSStyleSheet.js");
  if (isTarget && importer !== void 0 && normalizePath(importer).includes("react-native-web/dist/exports/StyleSheet")) {
    return styleSheetPath;
  }
};
const vite8OptimizeDeps = {
  exclude: ["uniwind", "react-native"],
  rolldownOptions: {
    plugins: [{
      name: "uniwind-rolldown-plugin",
      resolveId: resolveOrderedCSSStyleSheet
    }]
  }
};
const vite7OptimizeDeps = {
  exclude: ["uniwind", "react-native"],
  esbuildOptions: {
    plugins: [{
      name: "uniwind-esbuild-plugin",
      setup: (build) => {
        build.onResolve(
          { filter: /^\.\/createOrderedCSSStyleSheet$/ },
          (args) => {
            const resolved = resolveOrderedCSSStyleSheet(args.path, args.importer);
            if (resolved !== void 0) {
              return { path: resolved };
            }
          }
        );
      }
    }]
  }
};
const vite8Resolve = {
  alias: [{
    find: /^react-native$/,
    replacement: componentPath,
    customResolver: {
      resolveId(_, importer) {
        if (importer !== void 0 && normalizePath(importer).includes("uniwind/dist")) {
          return this.resolve("react-native-web", importer, { skipSelf: true });
        }
        return componentPath;
      }
    }
  }]
};
const vite7Resolve = {
  alias: [{
    find: /^react-native$/,
    replacement: componentPath,
    customResolver: {
      resolveId(_, importer) {
        if (importer !== void 0 && normalizePath(importer).includes("uniwind/dist")) {
          return this.resolve("react-native-web");
        }
        return componentPath;
      }
    }
  }]
};
const uniwind = (config) => {
  const bundlerConfig = UniwindBundlerConfig.fromViteConfig(config);
  return {
    name: "uniwind",
    enforce: "pre",
    resolveId: (source, importer) => {
      return resolveOrderedCSSStyleSheet(source, importer);
    },
    config: () => ({
      css: {
        transformer: "lightningcss",
        lightningcss: {
          visitor: bundlerConfig.cssVisitor
        }
      },
      optimizeDeps: isVite8 ? vite8OptimizeDeps : vite7OptimizeDeps,
      resolve: isVite8 ? vite8Resolve : vite7Resolve
    }),
    transform: (code, id) => {
      const normalizedId = normalizePath(id);
      if (normalizedId.includes("uniwind/dist") && normalizedId.includes("config/config.js")) {
        return {
          code: `${code}
;Uniwind.__reinit(() => ({}), ${bundlerConfig.stringifiedThemes})`
        };
      }
    },
    buildStart: async () => {
      await bundlerConfig.generateArtifacts(cssArtifactPath);
    },
    generateBundle: async () => {
      await bundlerConfig.generateArtifacts(cssArtifactPath);
    }
  };
};

export { uniwind };
