// uniwind@1.9.0/dist/vite/index.cjs
'use strict';

const node = require('@tailwindcss/node');
const node_module = require('node:module');
const path = require('path');
const config = require('../shared/uniwind.NUHr3In4.cjs');
require('fs');
require('lightningcss');

var _documentCurrentScript = typeof document !== 'undefined' ? document.currentScript : null;
function _interopDefaultCompat (e) { return e && typeof e === 'object' && 'default' in e ? e.default : e; }

const path__default = /*#__PURE__*/_interopDefaultCompat(path);

const dirname = typeof __dirname !== "undefined" ? __dirname : undefined;
const componentPath = path__default.resolve(
  dirname,
  "../module/components/web/index.js"
);
const styleSheetPath = path__default.resolve(
  dirname,
  "../module/components/web/createOrderedCSSStyleSheet.js"
);
const cssArtifactPath = path__default.resolve(dirname, "../../uniwind.css");
const require$1 = node_module.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.tagName.toUpperCase() === 'SCRIPT' && _documentCurrentScript.src || new URL('vite/index.cjs', document.baseURI).href)));
const viteVersion = require$1("vite/package.json").version;
const isVite8 = Number(viteVersion.split(".")[0]) >= 8;
const resolveOrderedCSSStyleSheet = (source, importer) => {
  const normalizedSource = node.normalizePath(source);
  const isTarget = source === "./createOrderedCSSStyleSheet" || normalizedSource.endsWith("react-native-web/dist/exports/StyleSheet/dom/createOrderedCSSStyleSheet.js");
  if (isTarget && importer !== void 0 && node.normalizePath(importer).includes("react-native-web/dist/exports/StyleSheet")) {
    return styleSheetPath;
  }
};
const vite8OptimizeDeps = {
  exclude: ["uniwind", "react-native"],
  rolldownOptions: {
    plugins: [{
      name: "uniwind-rolldown-plugin",
      resolveId: resolveOrderedCSSStyleSheet
    }]
  }
};
const vite7OptimizeDeps = {
  exclude: ["uniwind", "react-native"],
  esbuildOptions: {
    plugins: [{
      name: "uniwind-esbuild-plugin",
      setup: (build) => {
        build.onResolve(
          { filter: /^\.\/createOrderedCSSStyleSheet$/ },
          (args) => {
            const resolved = resolveOrderedCSSStyleSheet(args.path, args.importer);
            if (resolved !== void 0) {
              return { path: resolved };
            }
          }
        );
      }
    }]
  }
};
const vite8Resolve = {
  alias: [{
    find: /^react-native$/,
    replacement: componentPath,
    customResolver: {
      resolveId(_, importer) {
        if (importer !== void 0 && node.normalizePath(importer).includes("uniwind/dist")) {
          return this.resolve("react-native-web", importer, { skipSelf: true });
        }
        return componentPath;
      }
    }
  }]
};
const vite7Resolve = {
  alias: [{
    find: /^react-native$/,
    replacement: componentPath,
    customResolver: {
      resolveId(_, importer) {
        if (importer !== void 0 && node.normalizePath(importer).includes("uniwind/dist")) {
          return this.resolve("react-native-web");
        }
        return componentPath;
      }
    }
  }]
};
const uniwind = (config$1) => {
  const bundlerConfig = config.UniwindBundlerConfig.fromViteConfig(config$1);
  return {
    name: "uniwind",
    enforce: "pre",
    resolveId: (source, importer) => {
      return resolveOrderedCSSStyleSheet(source, importer);
    },
    config: () => ({
      css: {
        transformer: "lightningcss",
        lightningcss: {
          visitor: bundlerConfig.cssVisitor
        }
      },
      optimizeDeps: isVite8 ? vite8OptimizeDeps : vite7OptimizeDeps,
      resolve: isVite8 ? vite8Resolve : vite7Resolve
    }),
    transform: (code, id) => {
      const normalizedId = node.normalizePath(id);
      if (normalizedId.includes("uniwind/dist") && normalizedId.includes("config/config.js")) {
        return {
          code: `${code}
;Uniwind.__reinit(() => ({}), ${bundlerConfig.stringifiedThemes})`
        };
      }
    },
    buildStart: async () => {
      await bundlerConfig.generateArtifacts(cssArtifactPath);
    },
    generateBundle: async () => {
      await bundlerConfig.generateArtifacts(cssArtifactPath);
    }
  };
};

exports.uniwind = uniwind;
