// uniwind@1.9.0/dist/module/common/consts.js
export var Platform = /* @__PURE__ */ ((Platform2) => {
  Platform2["Android"] = "android";
  Platform2["iOS"] = "ios";
  Platform2["Web"] = "web";
  Platform2["Native"] = "native";
  Platform2["TV"] = "tv";
  Platform2["AndroidTV"] = "android-tv";
  Platform2["AppleTV"] = "apple-tv";
  return Platform2;
})(Platform || {});
export const UNIWIND_PLATFORM_VARIABLES = "__uniwind-platform-";
export const UNIWIND_THEME_VARIABLES = "__uniwind-theme-";
export var StyleDependency = /* @__PURE__ */ ((StyleDependency2) => {
  StyleDependency2[StyleDependency2["ColorScheme"] = 1] = "ColorScheme";
  StyleDependency2[StyleDependency2["Theme"] = 2] = "Theme";
  StyleDependency2[StyleDependency2["Dimensions"] = 3] = "Dimensions";
  StyleDependency2[StyleDependency2["Orientation"] = 4] = "Orientation";
  StyleDependency2[StyleDependency2["Insets"] = 5] = "Insets";
  StyleDependency2[StyleDependency2["FontScale"] = 6] = "FontScale";
  StyleDependency2[StyleDependency2["Rtl"] = 7] = "Rtl";
  StyleDependency2[StyleDependency2["AdaptiveThemes"] = 8] = "AdaptiveThemes";
  StyleDependency2[StyleDependency2["Variables"] = 9] = "Variables";
  return StyleDependency2;
})(StyleDependency || {});
export var Orientation = /* @__PURE__ */ ((Orientation2) => {
  Orientation2["Portrait"] = "portrait";
  Orientation2["Landscape"] = "landscape";
  return Orientation2;
})(Orientation || {});
export var ColorScheme = /* @__PURE__ */ ((ColorScheme2) => {
  ColorScheme2["Light"] = "light";
  ColorScheme2["Dark"] = "dark";
  return ColorScheme2;
})(ColorScheme || {});
