// uniwind@1.9.0/dist/module/components/native/RefreshControl.js
import { jsx } from "react/jsx-runtime";
import { RefreshControl as RNRefreshControl } from "react-native";
import { copyComponentProperties } from "../utils.js";
import { useAccentColor } from "./useAccentColor.js";
import { useStyle } from "./useStyle.js";
export const RefreshControl = copyComponentProperties(RNRefreshControl, (props) => {
  const style = useStyle(props.className, props);
  const color = useAccentColor(props.colorsClassName, props);
  const tintColor = useAccentColor(props.tintColorClassName, props);
  const titleColor = useAccentColor(props.titleColorClassName, props);
  const progressBackgroundColor = useAccentColor(props.progressBackgroundColorClassName, props);
  return /* @__PURE__ */ jsx(
    RNRefreshControl,
    {
      ...props,
      style: [style, props.style],
      colors: props.colors ?? (color !== void 0 ? [color] : void 0),
      tintColor: props.tintColor ?? tintColor,
      titleColor: props.titleColor ?? titleColor,
      progressBackgroundColor: props.progressBackgroundColor ?? progressBackgroundColor
    }
  );
});
export default RefreshControl;
