// uniwind@1.9.0/dist/module/components/native/Switch.js
import { jsx } from "react/jsx-runtime";
import { Switch as RNSwitch } from "react-native";
import { copyComponentProperties } from "../utils.js";
import { useAccentColor } from "./useAccentColor.js";
import { useStyle } from "./useStyle.js";
export const Switch = copyComponentProperties(RNSwitch, (props) => {
  const state = {
    isDisabled: Boolean(props.disabled)
  };
  const style = useStyle(props.className, props, state);
  const trackColorOn = useAccentColor(props.trackColorOnClassName, props, state);
  const trackColorOff = useAccentColor(props.trackColorOffClassName, props, state);
  const thumbColor = useAccentColor(props.thumbColorClassName, props, state);
  const ios_backgroundColor = useAccentColor(props.ios_backgroundColorClassName, props, state);
  return /* @__PURE__ */ jsx(
    RNSwitch,
    {
      ...props,
      style: [style, props.style],
      thumbColor: props.thumbColor ?? thumbColor,
      trackColor: { true: props.trackColor?.true ?? trackColorOn, false: props.trackColor?.false ?? trackColorOff },
      ios_backgroundColor: props.ios_backgroundColor ?? ios_backgroundColor
    }
  );
});
export default Switch;
