// uniwind@1.9.0/dist/module/components/native/TouchableHighlight.js
import { jsx } from "react/jsx-runtime";
import { useState } from "react";
import { TouchableHighlight as RNTouchableHighlight } from "react-native";
import { copyComponentProperties } from "../utils.js";
import { useAccentColor } from "./useAccentColor.js";
import { useStyle } from "./useStyle.js";
export const TouchableHighlight = copyComponentProperties(RNTouchableHighlight, (props) => {
  const [isPressed, setIsPressed] = useState(false);
  const [isFocused, setIsFocused] = useState(false);
  const state = {
    isDisabled: Boolean(props.disabled),
    isPressed,
    isFocused
  };
  const style = useStyle(props.className, props, state);
  const underlayColor = useAccentColor(props.underlayColorClassName, props, state);
  return /* @__PURE__ */ jsx(
    RNTouchableHighlight,
    {
      ...props,
      style: [style, props.style],
      underlayColor: props.underlayColor ?? underlayColor,
      onPressIn: (event) => {
        setIsPressed(true);
        props.onPressIn?.(event);
      },
      onPressOut: (event) => {
        setIsPressed(false);
        props.onPressOut?.(event);
      },
      onFocus: (event) => {
        setIsFocused(true);
        props.onFocus?.(event);
      },
      onBlur: (event) => {
        setIsFocused(false);
        props.onBlur?.(event);
      }
    }
  );
});
export default TouchableHighlight;
