// uniwind@1.9.0/dist/shared/uniwind.WHb-3yIn.mjs
import fs from 'fs';
import { compile } from '@tailwindcss/node';
import { transform } from 'lightningcss';
import path from 'path';

const toKebabCase = (str) => str.replace(/[A-Z]/g, (letter) => `-${letter.toLowerCase()}`);
const generateExtraUtilities = (map) => {
  return Object.entries(map).map(
    ([name, style]) => [
      `@utility ${name} {`,
      ...Object.entries(style).map(([key, value]) => `  ${toKebabCase(key)}: ${value};`),
      `}`,
      ""
    ].join("\n")
  ).join("\n");
};
const EXTRA_UTILITIES_MAP = {
  "border-continuous": {
    borderCurve: "continuous"
  },
  "border-circular": {
    borderCurve: "circular"
  }
};
const EXTRA_UTILITIES_CSS = generateExtraUtilities(EXTRA_UTILITIES_MAP);

const types = ["margin", "padding", "inset"];
const sides = ["inset", "x", "y", "top", "bottom", "left", "right"];
const safeAreaTypes = ["safe", "safe-or-*", "safe-offset-*"];
const spacing = "--spacing(--value(integer))";
const length = "--value([length], --spacing-*)";
const generateCSSForInsets = () => {
  let css = `@utility h-screen-safe {
    height: calc(100vh - (env(safe-area-inset-top) + env(safe-area-inset-bottom)));
}

`;
  const getInsetsForSide = (side) => {
    switch (side) {
      case "top":
        return ["top"];
      case "bottom":
        return ["bottom"];
      case "left":
        return ["left"];
      case "right":
        return ["right"];
      case "x":
        return ["left", "right"];
      case "y":
        return ["top", "bottom"];
      case "inset":
        return ["top", "bottom", "left", "right"];
      default:
        return [];
    }
  };
  const getUtilityName = (typeName, side, safeAreaType) => {
    if (typeName === "inset") {
      return `${side}-${safeAreaType}`;
    }
    const sideSuffix = side === "inset" ? "" : side.at(0);
    return `${typeName.at(0)}${sideSuffix}-${safeAreaType}`;
  };
  const getStyleProperty = (typeName, inset) => {
    if (typeName === "inset") {
      return inset;
    }
    return `${typeName}-${inset}`;
  };
  const getStylesForSafeAreaType = (safeAreaType, styles) => {
    switch (safeAreaType) {
      case "safe":
        return styles;
      case "safe-or-*":
        return styles.flatMap((style) => {
          const styleWithoutSemicolon = style.replace(";", "");
          return [
            styleWithoutSemicolon.replace(/: (env.*)/, (_, env) => `: max(${env}, ${spacing});`),
            styleWithoutSemicolon.replace(/: (env.*)/, (_, env) => `: max(${env}, ${length});`)
          ];
        });
      case "safe-offset-*":
        return styles.flatMap((style) => {
          const styleWithoutSemicolon = style.replace(";", "");
          return [
            styleWithoutSemicolon.replace(/: (env.*)/, (_, env) => `: calc(${env} + ${spacing});`),
            styleWithoutSemicolon.replace(/: (env.*)/, (_, env) => `: calc(${env} + ${length});`)
          ];
        });
      default:
        return [];
    }
  };
  types.forEach((type) => {
    sides.forEach((side) => {
      const insets = getInsetsForSide(side);
      const styles = insets.map((inset) => `${getStyleProperty(type, inset)}: env(safe-area-inset-${inset});`);
      safeAreaTypes.forEach((safeAreaType) => {
        const utilityName = getUtilityName(type, side, safeAreaType);
        css += [
          `@utility ${utilityName} {`,
          ...getStylesForSafeAreaType(safeAreaType, styles).map((style) => `    ${style}`),
          "}",
          "",
          ""
        ].join("\n");
      });
    });
  });
  return css.slice(0, -1);
};
const INSETS_CSS = generateCSSForInsets();

const overwriteDisabled = `@custom-variant disabled {
    &:disabled {
        @slot;
    }

    &[aria-disabled="true"] {
        @slot;
    }

    &[readonly] {
        @slot;
    }
}
`;
const OVERWRITE_CSS = overwriteDisabled;

const red = "\x1B[91m";
const blue = "\x1B[36m";
const reset = "\x1B[0m";
class Logger {
  constructor(name) {
    this.name = name;
  }
  static debug = false;
  static log(message, meta = "") {
    {
      return;
    }
  }
  static info(message) {
    console.log(`${blue}[Uniwind] ${message}${reset}`);
  }
  static error(message, meta = "") {
    console.log(`${red}Uniwind Error ${meta}- ${message}${reset}`);
  }
  static warn(message, meta = "") {
    {
      return;
    }
  }
  log(message) {
    Logger.log(message, `[${this.name} Processor] `);
  }
  error(message) {
    Logger.error(message, `[${this.name} Processor] `);
  }
  warn(message) {
    Logger.warn(message, `[${this.name} Processor] `);
  }
}

const readFileSafe = (filePath) => {
  try {
    return fs.readFileSync(filePath, "utf-8");
  } catch {
    return null;
  }
};
const isExcludedDependency = (url) => [
  url.includes("node_modules/tailwindcss"),
  url.includes("node_modules/@tailwindcss"),
  url.includes("node_modules/uniwind")
].some(Boolean);
const generateCSSForThemes = async (themes, input) => {
  const themesVariables = Object.fromEntries(themes.map((theme) => [theme, /* @__PURE__ */ new Set()]));
  const findVariantsRec = async (cssPath) => {
    const css = readFileSafe(cssPath);
    if (css === null) {
      return;
    }
    const { dependencies } = transform({
      code: Buffer.from(css),
      filename: "uniwind.css",
      analyzeDependencies: true,
      visitor: {
        Rule: (rule) => {
          if (rule.type === "unknown" && rule.value.name === "variant") {
            const [firstPrelude] = rule.value.prelude;
            if (firstPrelude?.type !== "token" || firstPrelude.value.type !== "ident" || !themes.includes(firstPrelude.value.value)) {
              return;
            }
            const theme = firstPrelude.value.value;
            rule.value.block?.forEach((block) => {
              if (block.type === "dashed-ident") {
                themesVariables[theme]?.add(block.value);
              }
            });
          }
        }
      }
    });
    if (!Array.isArray(dependencies)) {
      return;
    }
    const importUrls = /* @__PURE__ */ new Set();
    const importsCSS = dependencies.filter((dependency) => {
      if (dependency.type !== "import") {
        return false;
      }
      if (dependency.url.startsWith(".")) {
        importUrls.add(path.resolve(path.dirname(cssPath), dependency.url));
        return false;
      }
      return !isExcludedDependency(dependency.url);
    }).map((dependency) => `@import "${dependency.url}";`).join("\n");
    await compile(importsCSS, {
      base: path.resolve(path.dirname(cssPath)),
      onDependency: (dependency) => {
        if (isExcludedDependency(dependency)) {
          return;
        }
        importUrls.add(dependency);
      }
    });
    for (const filePath of importUrls) {
      await findVariantsRec(filePath);
    }
  };
  await findVariantsRec(input);
  let hasErrors = false;
  const hasVariables = Object.values(themesVariables).some((variables) => variables.size > 0);
  Object.values(themesVariables).forEach((variables) => {
    Object.entries(themesVariables).forEach(([checkedTheme, checkedVariables]) => {
      variables.forEach((variable) => {
        if (!checkedVariables.has(variable)) {
          Logger.error(`Theme ${checkedTheme} is missing variable ${variable}`);
          hasErrors = true;
        }
      });
    });
  });
  if (hasErrors) {
    Logger.error("All themes must have the same variables");
  }
  const variablesCSS = hasVariables ? [
    "",
    "@theme {",
    ...Array.from(Object.values(themesVariables).at(0) ?? []).map((variable) => `    ${variable}: unset;`),
    "}"
  ] : [];
  const uniwindCSS = [
    ...themes.map((theme) => {
      const notOtherThemes = themes.map((t) => `.${t}, .${t} *`);
      if (theme === "dark" || theme === "light") {
        return [
          `@custom-variant ${theme} {`,
          `   &:where(.${theme}, .${theme} *) {`,
          "       @slot;",
          "   }",
          "",
          `   @media (prefers-color-scheme: ${theme}) {`,
          `       &:not(:where(${notOtherThemes.join(", ")})) {`,
          "           @slot;",
          "       }",
          "   }",
          "}",
          ""
        ].join("\n");
      }
      return `@custom-variant ${theme} (&:where(.${theme}, .${theme} *));`;
    }),
    ...variablesCSS
  ].join("\n");
  return uniwindCSS;
};

const variants = ["ios", "android", "web", "native", "tv", "android-tv", "apple-tv"];
const generateCSSForVariants = () => {
  let css = "";
  variants.forEach((variant) => {
    css += `@custom-variant ${variant} (${variant === "web" ? "@supports selector(div > div)" : `@media ${variant}`});
`;
  });
  return css;
};
const VARIANTS_CSS = generateCSSForVariants();

const buildCSS = async (themes, input, cssFilePath) => {
  const themesCSS = await generateCSSForThemes(themes, input);
  const oldCSSFile = fs.existsSync(cssFilePath) ? fs.readFileSync(cssFilePath, "utf-8") : "";
  const newCssFile = [
    VARIANTS_CSS,
    INSETS_CSS,
    OVERWRITE_CSS,
    EXTRA_UTILITIES_CSS,
    themesCSS
  ].join("\n");
  if (oldCSSFile === newCssFile) {
    return;
  }
  fs.writeFileSync(
    cssFilePath,
    newCssFile
  );
};

const buildDtsFile = (dtsPath, stringifiedThemes) => {
  const oldDtsContent = fs.existsSync(dtsPath) ? fs.readFileSync(dtsPath, "utf-8") : "";
  const dtsContent = [
    `// NOTE: This file is generated by uniwind and it should not be edited manually.`,
    `/// <reference types="uniwind/types" />`,
    "",
    `declare module 'uniwind' {`,
    `    export interface UniwindConfig {`,
    `        themes: readonly ${stringifiedThemes}`,
    `    }`,
    `}`,
    "",
    `export {}`,
    ""
  ].join("\n");
  if (oldDtsContent === dtsContent) {
    return;
  }
  fs.writeFileSync(dtsPath, dtsContent);
};

const ONE_PX = {
  type: "token",
  value: { type: "dimension", unit: "px", value: 1 }
};
class FunctionVisitor {
  pixelRatio(fn) {
    return {
      type: "function",
      value: {
        name: "calc",
        arguments: [
          fn.arguments.at(0) ?? ONE_PX,
          { type: "token", value: { type: "delim", value: "*" } },
          ONE_PX
        ]
      }
    };
  }
  fontScale(fn) {
    return {
      type: "function",
      value: {
        name: "calc",
        arguments: [
          fn.arguments.at(0) ?? ONE_PX,
          { type: "token", value: { type: "delim", value: "*" } },
          {
            type: "token",
            value: { type: "dimension", value: 1, unit: "rem" }
          }
        ]
      }
    };
  }
  hairlineWidth() {
    return ONE_PX;
  }
}

class RuleVisitor {
  constructor(bundlerConfig) {
    this.bundlerConfig = bundlerConfig;
  }
  processedClassNames = /* @__PURE__ */ new Set();
  processedVariables = /* @__PURE__ */ new Set();
  currentLayerName = "";
  "layer-block" = (layer) => {
    this.currentLayerName = layer.value.name?.join("") ?? "";
  };
  style = (styleRule) => {
    const firstSelector = styleRule.value.selectors.at(0)?.at(0);
    if (this.currentLayerName === "theme" && firstSelector?.type === "pseudo-class" && firstSelector.kind === "root") {
      return this.removeNulls(this.processThemeRoot(styleRule));
    }
    if (firstSelector?.type === "class") {
      return this.processClassStyle(styleRule, firstSelector);
    }
  };
  cleanup() {
    this.currentLayerName = "";
    this.processedClassNames.clear();
    this.processedVariables.clear();
  }
  processThemeRoot(styleRule) {
    const themeScopedRules = styleRule.value.rules?.filter((rule) => {
      if (rule.type !== "style") {
        return false;
      }
      const firstSelector = rule.value.selectors.at(0)?.at(0);
      const secondSelector = rule.value.selectors.at(0)?.at(1);
      return firstSelector?.type === "nesting" && secondSelector?.type === "pseudo-class" && secondSelector.kind === "where";
    }) ?? [];
    const nonThemeRules = styleRule.value.rules?.filter((rule) => !themeScopedRules.includes(rule));
    const processedThemeScopedRules = themeScopedRules.map((rule) => {
      if (rule.type !== "style") {
        return rule;
      }
      const secondSelector = rule.value.selectors.at(0)?.at(1);
      if (secondSelector?.type === "pseudo-class" && secondSelector.kind === "where") {
        return this.processThemeStyle(rule, secondSelector);
      }
      return rule;
    });
    return [
      {
        type: "style",
        value: {
          loc: styleRule.value.loc,
          selectors: styleRule.value.selectors,
          rules: nonThemeRules,
          declarations: styleRule.value.declarations
        }
      },
      ...processedThemeScopedRules
    ];
  }
  processThemeStyle(styleRule, secondSelector) {
    const whereSelector = secondSelector.selectors.at(0)?.at(0);
    if (whereSelector?.type !== "class") {
      return styleRule;
    }
    const selectedVariant = this.bundlerConfig.themes.find((theme) => whereSelector.name === theme);
    if (selectedVariant === void 0 || this.processedVariables.has(selectedVariant)) {
      return styleRule;
    }
    this.processedVariables.add(selectedVariant);
    return {
      type: "style",
      value: {
        loc: styleRule.value.loc,
        selectors: [[{ type: "class", name: selectedVariant }]],
        declarations: styleRule.value.declarations,
        rules: styleRule.value.rules
      }
    };
  }
  processClassStyle(styleRule, firstSelector) {
    const selectedVariant = this.bundlerConfig.themes.find((theme) => firstSelector.name.includes(`${theme}:`));
    if (selectedVariant === void 0 || this.processedClassNames.has(firstSelector.name)) {
      return;
    }
    this.processedClassNames.add(firstSelector.name);
    return this.removeNulls({
      type: "scope",
      value: {
        loc: styleRule.value.loc,
        rules: [styleRule],
        scopeStart: [[{ type: "class", name: selectedVariant }]],
        scopeEnd: this.bundlerConfig.themes.filter((theme) => theme !== selectedVariant).map((theme) => [{ type: "class", name: theme }])
      }
    });
  }
  // Fixes lightningcss serialization bug
  removeNulls(value) {
    if (Array.isArray(value)) {
      return value.map((v) => this.removeNulls(v));
    }
    if (typeof value === "object" && value !== null) {
      return Object.fromEntries(
        Object.entries(value).filter(([_, value2]) => {
          if (value2 === null) {
            return false;
          }
          return true;
        }).map(([key, value2]) => [key, this.removeNulls(value2)])
      );
    }
    return value;
  }
}

class UniwindCSSVisitor {
  Function;
  Rule;
  StyleSheet;
  constructor(bundlerConfig) {
    const ruleVisitor = new RuleVisitor(bundlerConfig);
    this.Function = new FunctionVisitor();
    this.Rule = ruleVisitor;
    this.StyleSheet = () => {
      ruleVisitor.cleanup();
    };
  }
}

var Platform = /* @__PURE__ */ ((Platform2) => {
  Platform2["Android"] = "android";
  Platform2["iOS"] = "ios";
  Platform2["Web"] = "web";
  Platform2["Native"] = "native";
  Platform2["TV"] = "tv";
  Platform2["AndroidTV"] = "android-tv";
  Platform2["AppleTV"] = "apple-tv";
  return Platform2;
})(Platform || {});
const UNIWIND_PLATFORM_VARIABLES = "__uniwind-platform-";
const UNIWIND_THEME_VARIABLES = "__uniwind-theme-";
var StyleDependency = /* @__PURE__ */ ((StyleDependency2) => {
  StyleDependency2[StyleDependency2["ColorScheme"] = 1] = "ColorScheme";
  StyleDependency2[StyleDependency2["Theme"] = 2] = "Theme";
  StyleDependency2[StyleDependency2["Dimensions"] = 3] = "Dimensions";
  StyleDependency2[StyleDependency2["Orientation"] = 4] = "Orientation";
  StyleDependency2[StyleDependency2["Insets"] = 5] = "Insets";
  StyleDependency2[StyleDependency2["FontScale"] = 6] = "FontScale";
  StyleDependency2[StyleDependency2["Rtl"] = 7] = "Rtl";
  StyleDependency2[StyleDependency2["AdaptiveThemes"] = 8] = "AdaptiveThemes";
  StyleDependency2[StyleDependency2["Variables"] = 9] = "Variables";
  return StyleDependency2;
})(StyleDependency || {});

const isDefined = (value) => value !== void 0 && value !== null;

class UniwindBundlerConfig {
  constructor(config, platform) {
    this.config = config;
    this.platform = platform;
  }
  static fromMetroConfig(config, platform) {
    const getPlatform = () => {
      if (!isDefined(platform)) {
        return Platform.Native;
      }
      if (!config.isTV) {
        return platform;
      }
      if (platform === Platform.Android) {
        return Platform.AndroidTV;
      }
      if (platform === Platform.iOS) {
        return Platform.AppleTV;
      }
      throw new Error(`Platform ${platform} not supported`);
    };
    if (typeof config === "undefined") {
      throw new Error("Uniwind: You need to pass second parameter to withUniwindConfig");
    }
    if (typeof config.cssEntryFile === "undefined") {
      throw new Error(
        'Uniwind: You need to pass css css entry file to withUniwindConfig, e.g. withUniwindConfig(config, { cssEntryFile: "./global.css" })'
      );
    }
    return new UniwindBundlerConfig(config, getPlatform());
  }
  static fromViteConfig(config) {
    return new UniwindBundlerConfig(config, Platform.Web);
  }
  static fromCliConfig(config) {
    if (typeof config.cssEntryFile === "undefined") {
      throw new Error(
        "Uniwind: You need to pass css entry file, e.g. uniwind generate-artifacts --css ./global.css. Run uniwind generate-artifacts --help for usage."
      );
    }
    return new UniwindBundlerConfig(config, Platform.Web);
  }
  get cssPath() {
    return path.join(process.cwd(), this.config.cssEntryFile);
  }
  get themes() {
    return Array.from(
      /* @__PURE__ */ new Set([
        "light",
        "dark",
        ...this.config.extraThemes ?? []
      ])
    );
  }
  get cssVisitor() {
    return new UniwindCSSVisitor(this);
  }
  get polyfills() {
    return this.config.polyfills;
  }
  get stringifiedThemes() {
    return `[${this.themes.map((theme) => `'${theme}'`).join(", ")}]`;
  }
  toMetroConfig() {
    return this.config;
  }
  async generateArtifacts(cssArtifactPath) {
    await buildCSS(this.themes, this.config.cssEntryFile, cssArtifactPath);
    buildDtsFile(this.config.dtsFile ?? "uniwind-types.d.ts", this.stringifiedThemes);
  }
}

export { Logger as L, Platform as P, StyleDependency as S, UniwindBundlerConfig as U, UNIWIND_PLATFORM_VARIABLES as a, UNIWIND_THEME_VARIABLES as b, isDefined as i };
