// uniwind@1.9.0/dist/common/bundler/css-visitor/function-visitor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FunctionVisitor = void 0;
const ONE_PX = {
  type: "token",
  value: {
    type: "dimension",
    unit: "px",
    value: 1
  }
};
class FunctionVisitor {
  pixelRatio(fn) {
    return {
      type: "function",
      value: {
        name: "calc",
        arguments: [fn.arguments.at(0) ?? ONE_PX, {
          type: "token",
          value: {
            type: "delim",
            value: "*"
          }
        }, ONE_PX]
      }
    };
  }
  fontScale(fn) {
    return {
      type: "function",
      value: {
        name: "calc",
        arguments: [fn.arguments.at(0) ?? ONE_PX, {
          type: "token",
          value: {
            type: "delim",
            value: "*"
          }
        }, {
          type: "token",
          value: {
            type: "dimension",
            value: 1,
            unit: "rem"
          }
        }]
      }
    };
  }
  hairlineWidth() {
    return ONE_PX;
  }
}
exports.FunctionVisitor = FunctionVisitor;
