// uniwind@1.9.0/dist/common/bundler/css-visitor/rule-visitor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RuleVisitor = void 0;
class RuleVisitor {
  constructor(bundlerConfig) {
    this.bundlerConfig = bundlerConfig;
  }
  processedClassNames = /* @__PURE__ */new Set();
  processedVariables = /* @__PURE__ */new Set();
  currentLayerName = "";
  "layer-block" = layer => {
    this.currentLayerName = layer.value.name?.join("") ?? "";
  };
  style = styleRule => {
    const firstSelector = styleRule.value.selectors.at(0)?.at(0);
    if (this.currentLayerName === "theme" && firstSelector?.type === "pseudo-class" && firstSelector.kind === "root") {
      return this.removeNulls(this.processThemeRoot(styleRule));
    }
    if (firstSelector?.type === "class") {
      return this.processClassStyle(styleRule, firstSelector);
    }
  };
  cleanup() {
    this.currentLayerName = "";
    this.processedClassNames.clear();
    this.processedVariables.clear();
  }
  processThemeRoot(styleRule) {
    const themeScopedRules = styleRule.value.rules?.filter(rule => {
      if (rule.type !== "style") {
        return false;
      }
      const firstSelector = rule.value.selectors.at(0)?.at(0);
      const secondSelector = rule.value.selectors.at(0)?.at(1);
      return firstSelector?.type === "nesting" && secondSelector?.type === "pseudo-class" && secondSelector.kind === "where";
    }) ?? [];
    const nonThemeRules = styleRule.value.rules?.filter(rule => !themeScopedRules.includes(rule));
    const processedThemeScopedRules = themeScopedRules.map(rule => {
      if (rule.type !== "style") {
        return rule;
      }
      const secondSelector = rule.value.selectors.at(0)?.at(1);
      if (secondSelector?.type === "pseudo-class" && secondSelector.kind === "where") {
        return this.processThemeStyle(rule, secondSelector);
      }
      return rule;
    });
    return [{
      type: "style",
      value: {
        loc: styleRule.value.loc,
        selectors: styleRule.value.selectors,
        rules: nonThemeRules,
        declarations: styleRule.value.declarations
      }
    }, ...processedThemeScopedRules];
  }
  processThemeStyle(styleRule, secondSelector) {
    const whereSelector = secondSelector.selectors.at(0)?.at(0);
    if (whereSelector?.type !== "class") {
      return styleRule;
    }
    const selectedVariant = this.bundlerConfig.themes.find(theme => whereSelector.name === theme);
    if (selectedVariant === void 0 || this.processedVariables.has(selectedVariant)) {
      return styleRule;
    }
    this.processedVariables.add(selectedVariant);
    return {
      type: "style",
      value: {
        loc: styleRule.value.loc,
        selectors: [[{
          type: "class",
          name: selectedVariant
        }]],
        declarations: styleRule.value.declarations,
        rules: styleRule.value.rules
      }
    };
  }
  processClassStyle(styleRule, firstSelector) {
    const selectedVariant = this.bundlerConfig.themes.find(theme => firstSelector.name.includes(`${theme}:`));
    if (selectedVariant === void 0 || this.processedClassNames.has(firstSelector.name)) {
      return;
    }
    this.processedClassNames.add(firstSelector.name);
    return this.removeNulls({
      type: "scope",
      value: {
        loc: styleRule.value.loc,
        rules: [styleRule],
        scopeStart: [[{
          type: "class",
          name: selectedVariant
        }]],
        scopeEnd: this.bundlerConfig.themes.filter(theme => theme !== selectedVariant).map(theme => [{
          type: "class",
          name: theme
        }])
      }
    });
  }
  // Fixes lightningcss serialization bug
  removeNulls(value) {
    if (Array.isArray(value)) {
      return value.map(v => this.removeNulls(v));
    }
    if (typeof value === "object" && value !== null) {
      return Object.fromEntries(Object.entries(value).filter(([_, value2]) => {
        if (value2 === null) {
          return false;
        }
        return true;
      }).map(([key, value2]) => [key, this.removeNulls(value2)]));
    }
    return value;
  }
}
exports.RuleVisitor = RuleVisitor;
