// uniwind@1.9.0/dist/module/core/web/cssListener.js
import { StyleDependency } from "../../common/consts.js";
import { UniwindListener } from "../listener.js";
class CSSListenerBuilder {
  activeRules = /* @__PURE__ */ new Set();
  classNameMediaQueryListeners = /* @__PURE__ */ new Map();
  listeners = /* @__PURE__ */ new Map();
  registeredRulesMediaQueries = /* @__PURE__ */ new Map();
  processedStyleSheets = /* @__PURE__ */ new WeakSet();
  pendingInitialization = void 0;
  constructor() {
    if (typeof document === "undefined") {
      return;
    }
    const observer = new MutationObserver((mutations) => {
      for (const mutation of mutations) {
        if (mutation.type === "attributes") {
          const el = mutation.target;
          if (!("sheet" in el)) {
            continue;
          }
          const sheet = el.sheet;
          if (sheet) {
            this.processedStyleSheets.delete(sheet);
          }
          this.scheduleInitialization();
        }
        if (mutation.type === "childList") {
          this.scheduleInitialization();
        }
      }
    });
    this.initialize();
    observer.observe(document.head, {
      childList: true,
      subtree: false,
      attributes: true,
      attributeFilter: ["disabled", "media", "title", "href", "rel"]
    });
  }
  subscribeToClassName(classNames, listener) {
    const disposables = [];
    classNames.split(" ").forEach((className) => {
      const mediaQuery = this.classNameMediaQueryListeners.get(className);
      if (!mediaQuery) {
        return () => {
        };
      }
      const listeners = this.listeners.get(mediaQuery);
      listeners?.add(listener);
      disposables.push(() => listeners?.delete(listener));
    });
    const disposeThemeListener = UniwindListener.subscribe(listener, [StyleDependency.Theme, StyleDependency.Variables]);
    return () => {
      disposables.forEach((disposable) => disposable());
      disposeThemeListener();
    };
  }
  scheduleInitialization() {
    this.cancelPendingInitialization();
    if (typeof requestIdleCallback !== "undefined") {
      this.pendingInitialization = requestIdleCallback(() => {
        this.initialize();
      }, { timeout: 50 });
      return;
    }
    this.pendingInitialization = setTimeout(() => {
      this.initialize();
    }, 50);
  }
  cancelPendingInitialization() {
    if (this.pendingInitialization !== void 0) {
      if (typeof cancelIdleCallback !== "undefined") {
        cancelIdleCallback(this.pendingInitialization);
      } else {
        clearTimeout(this.pendingInitialization);
      }
      this.pendingInitialization = void 0;
    }
  }
  pruneStaleRules() {
    const activeSheets = new Set(Array.from(document.styleSheets));
    for (const rule of this.activeRules) {
      if (!rule.parentStyleSheet || !activeSheets.has(rule.parentStyleSheet)) {
        this.activeRules.delete(rule);
      }
    }
  }
  initialize() {
    this.pendingInitialization = void 0;
    this.pruneStaleRules();
    for (const sheet of Array.from(document.styleSheets)) {
      if (this.processedStyleSheets.has(sheet)) {
        continue;
      }
      let rules;
      try {
        rules = sheet.cssRules;
      } catch {
        continue;
      }
      if (!rules) {
        continue;
      }
      this.processedStyleSheets.add(sheet);
      this.addMediaQueriesDeep(rules);
    }
  }
  isStyleRule(rule) {
    return rule.constructor.name === "CSSStyleRule";
  }
  isMediaRule(rule) {
    return rule.constructor.name === "CSSMediaRule";
  }
  isSupportsRule(rule) {
    return rule.constructor.name === "CSSSupportsRule";
  }
  collectParentMediaQueries(rule, acc = []) {
    const { parentRule } = rule;
    if (!parentRule) {
      return [];
    }
    if (this.isMediaRule(parentRule)) {
      acc.push(parentRule);
    }
    const result = this.collectParentMediaQueries(parentRule, acc);
    acc.push(...result);
    return Array.from(new Set(acc));
  }
  addMediaQueriesDeep(rules) {
    for (const rule of Array.from(rules)) {
      if (this.isStyleRule(rule)) {
        const mediaQueries = this.collectParentMediaQueries(rule);
        this.activeRules.add(rule);
        if (mediaQueries.length > 0) {
          this.addMediaQuery(mediaQueries, rule);
        }
        continue;
      }
      if (this.isSupportsRule(rule)) {
        if (!CSS.supports(rule.conditionText)) {
          continue;
        }
        this.addMediaQueriesDeep(rule.cssRules);
        continue;
      }
      if ("cssRules" in rule && rule.cssRules instanceof CSSRuleList) {
        this.addMediaQueriesDeep(rule.cssRules);
        continue;
      }
    }
  }
  addMediaQuery(mediaQueries, rule) {
    const className = rule.selectorText;
    const rules = mediaQueries.map((mediaQuery) => mediaQuery.conditionText).sort().join(" and ");
    const parsedClassName = className.replace(".", "").replace("\\", "");
    const cachedMediaQueryList = this.registeredRulesMediaQueries.get(rules);
    if (cachedMediaQueryList) {
      this.classNameMediaQueryListeners.set(parsedClassName, cachedMediaQueryList);
      this.toggleRule(cachedMediaQueryList, rule);
      cachedMediaQueryList.addEventListener("change", () => {
        this.toggleRule(cachedMediaQueryList, rule);
      });
      return;
    }
    const mediaQueryList = window.matchMedia(rules);
    this.toggleRule(mediaQueryList, rule);
    this.registeredRulesMediaQueries.set(rules, mediaQueryList);
    this.listeners.set(mediaQueryList, /* @__PURE__ */ new Set());
    this.classNameMediaQueryListeners.set(parsedClassName, mediaQueryList);
    mediaQueryList.addEventListener("change", () => {
      this.listeners.get(mediaQueryList).forEach((listener) => {
        listener();
      });
      this.toggleRule(mediaQueryList, rule);
    });
  }
  isRuleLive(rule) {
    const sheet = rule.parentStyleSheet;
    return sheet !== null && Array.from(document.styleSheets).includes(sheet);
  }
  toggleRule(mqList, rule) {
    if (mqList.matches && this.isRuleLive(rule)) {
      this.activeRules.add(rule);
    } else {
      this.activeRules.delete(rule);
    }
  }
}
export const CSSListener = new CSSListenerBuilder();
