// uniwind@1.9.0/dist/module/bundler/css-processor/mq.js
import { Platform } from "@/common/consts";
export class MQ {
  constructor(Processor) {
    this.Processor = Processor;
  }
  processMediaQueries(mediaQueries) {
    const mq = this.getInitialMediaQueryResolver();
    mediaQueries.forEach((mediaQuery) => {
      const { condition, mediaType } = mediaQuery;
      if ([
        Platform.Android,
        Platform.iOS,
        Platform.Native,
        Platform.AndroidTV,
        Platform.AppleTV,
        Platform.TV
      ].includes(mediaType)) {
        mq.platform = mediaType;
        return;
      }
      if (condition?.type !== "feature") {
        return;
      }
      if (condition.value.type === "range") {
        this.processWidthMediaQuery(condition.value, mq);
      }
      if (condition.value.type === "plain") {
        this.processPlainMediaQuery(condition.value, mq);
      }
    });
    return mq;
  }
  processWidthMediaQuery(query, mq) {
    const { operator, value } = query;
    const result = this.Processor.CSS.processValue(value);
    if (operator === "greater-than-equal" || operator === "greater-than") {
      mq.minWidth = result;
    }
    if (operator === "less-than-equal" || operator === "less-than") {
      mq.maxWidth = result;
    }
  }
  processPlainMediaQuery(query, mq) {
    const { value, name } = query;
    switch (name) {
      case "orientation":
        mq.orientation = value.value;
        break;
      case "prefers-color-scheme":
        mq.colorScheme = value.value;
        break;
      default:
        break;
    }
  }
  getInitialMediaQueryResolver() {
    return {
      minWidth: 0,
      maxWidth: Number.MAX_VALUE,
      platform: null,
      rtl: null,
      important: false,
      colorScheme: null,
      orientation: null,
      theme: null,
      active: null,
      focus: null,
      disabled: null,
      dataAttributes: null
    };
  }
}
