// uniwind@1.9.0/dist/metro/index.mjs
import { U as UniwindBundlerConfig, P as Platform } from '../shared/uniwind.WHb-3yIn.mjs';
import { dirname, sep, basename, join } from 'node:path';
import FileStoreBase from 'metro-cache/private/stores/FileStore';
import os from 'os';
import path from 'path';
import 'fs';
import '@tailwindcss/node';
import 'lightningcss';

class FileStore extends FileStoreBase {
  async set(key, value) {
    if (value?.output?.[0]?.data?.css?.skipCache) {
      return;
    }
    return super.set(key, value);
  }
}
const cacheStore = new FileStore({
  root: path.join(os.tmpdir(), "metro-cache")
});
const patchMetroGraphToSupportUncachedModules = () => {
  const { Graph } = require("metro/private/DeltaBundler/Graph");
  const original_traverseDependencies = Graph.prototype.traverseDependencies;
  if (original_traverseDependencies.__patched) {
    return;
  }
  original_traverseDependencies.__patched = true;
  function traverseDependencies(paths, options) {
    this.dependencies.forEach((dependency) => {
      if (dependency.output.find((file) => file.data?.css?.skipCache === true) && !paths.includes(dependency.path)) {
        dependency.unstable_transformResultKey = `${dependency.unstable_transformResultKey}.`;
        paths.push(dependency.path);
      }
    });
    return original_traverseDependencies.call(this, paths, options);
  }
  Graph.prototype.traverseDependencies = traverseDependencies;
  traverseDependencies.__patched = true;
};

let cachedInternalBasePath = null;
const SUPPORTED_COMPONENTS = [
  "ActivityIndicator",
  "Button",
  "FlatList",
  "Image",
  "ImageBackground",
  "InputAccessoryView",
  "KeyboardAvoidingView",
  "Modal",
  "Pressable",
  "RefreshControl",
  "SafeAreaView",
  "ScrollView",
  "SectionList",
  "Switch",
  "Text",
  "TextInput",
  "TouchableHighlight",
  "TouchableNativeFeedback",
  "TouchableOpacity",
  "TouchableWithoutFeedback",
  "View",
  "VirtualizedList",
  "createOrderedCSSStyleSheet"
];
const nativeResolver = ({
  context,
  moduleName,
  platform,
  resolver
}) => {
  const resolution = resolver(context, moduleName, platform);
  if (cachedInternalBasePath === null) {
    try {
      cachedInternalBasePath = dirname(require.resolve("uniwind/package.json"));
    } catch {
      cachedInternalBasePath = "";
    }
  }
  const isInternal = cachedInternalBasePath !== "" && context.originModulePath.startsWith(cachedInternalBasePath);
  const isFromNodeModules = context.originModulePath.includes(`${sep}node_modules${sep}`);
  const isFromReactNative = context.originModulePath.includes(`${sep}react-native${sep}`) || context.originModulePath.includes(`${sep}@react-native${sep}`);
  const isReactNativeAnimated = context.originModulePath.includes(`${sep}Animated${sep}components${sep}`);
  if (isInternal || resolution.type !== "sourceFile" || isFromReactNative && isFromNodeModules && !isReactNativeAnimated) {
    return resolution;
  }
  if (moduleName === "react-native") {
    return resolver(context, `uniwind/components`, platform);
  }
  if (resolution.filePath.includes(`${sep}react-native${sep}Libraries${sep}`)) {
    const filename = basename(resolution.filePath.split(sep).at(-1) ?? "");
    const module = filename.split(".").at(0);
    if (module !== void 0 && SUPPORTED_COMPONENTS.includes(module)) {
      return resolver(context, `uniwind/components/${module}`, platform);
    }
  }
  return resolution;
};
const webResolver = ({
  context,
  moduleName,
  platform,
  resolver
}) => {
  const resolution = resolver(context, moduleName, platform);
  if (cachedInternalBasePath === null) {
    try {
      cachedInternalBasePath = dirname(require.resolve("uniwind/package.json"));
    } catch {
      cachedInternalBasePath = "";
    }
  }
  if (cachedInternalBasePath !== "" && context.originModulePath.startsWith(cachedInternalBasePath) || resolution.type !== "sourceFile" || !resolution.filePath.includes(`${sep}react-native-web${sep}`)) {
    return resolution;
  }
  const segments = resolution.filePath.split(sep);
  const filename = segments.at(-1) ?? "";
  const isIndex = filename.startsWith("index.");
  const module = segments.at(-2);
  if (filename.startsWith("createOrderedCSSStyleSheet.")) {
    return resolver(context, `uniwind/components/createOrderedCSSStyleSheet`, platform);
  }
  if (!isIndex || module === void 0 || !SUPPORTED_COMPONENTS.includes(module) || context.originModulePath.endsWith(`${module}${sep}index.js`)) {
    return resolution;
  }
  return resolver(context, `uniwind/components/${module}`, platform);
};

const isUniwindRequest = (moduleName) => moduleName === "uniwind" || moduleName.startsWith("uniwind/");
const withUniwindConfig = (config, uniwindConfig) => {
  const bundlerConfig = UniwindBundlerConfig.fromMetroConfig(uniwindConfig);
  const pinnedUniwindOrigin = join(config.projectRoot ?? process.cwd(), "package.json");
  patchMetroGraphToSupportUncachedModules();
  return {
    ...config,
    cacheStores: [cacheStore],
    transformerPath: require.resolve("./transformer.cjs"),
    transformer: {
      ...config.transformer,
      uniwind: bundlerConfig.toMetroConfig()
    },
    resolver: {
      ...config.resolver,
      sourceExts: [
        ...config.resolver?.sourceExts ?? [],
        "css"
      ],
      assetExts: config.resolver?.assetExts?.filter(
        (ext) => ext !== "css"
      ),
      resolveRequest: (context, moduleName, platform) => {
        const baseResolver = config.resolver?.resolveRequest ?? context.resolveRequest;
        const resolver = (nextContext, nextModuleName, nextPlatform) => {
          if (isUniwindRequest(nextModuleName)) {
            return baseResolver(
              {
                ...nextContext,
                originModulePath: pinnedUniwindOrigin
              },
              nextModuleName,
              nextPlatform
            );
          }
          return baseResolver(nextContext, nextModuleName, nextPlatform);
        };
        const platformResolver = platform === Platform.Web ? webResolver : nativeResolver;
        const resolved = platformResolver({
          context,
          moduleName,
          platform,
          resolver
        });
        return resolved;
      }
    }
  };
};

export { withUniwindConfig };
