// uniwind@1.9.0/dist/common/hoc/withUniwind.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.withUniwind = void 0;
var _jsxRuntime = require("react/jsx-runtime");
var _react = require("react");
var _utils = require("../common/utils");
var _generateDataSet = require("../components/web/generateDataSet");
var _context = require("../core/context");
var _logger = require("../core/logger");
var _web = require("../core/web");
var _withUniwindUtils = require("./withUniwindUtils");
let warnedOnce = false;
const withUniwind = (Component, options) => options ? withManualUniwind(Component, options) : withAutoUniwind(Component);
exports.withUniwind = withUniwind;
const withAutoUniwind = Component => props => {
  const uniwindContext = (0, _context.useUniwindContext)();
  const {
    classNames,
    generatedProps
  } = Object.entries(props).reduce((acc, [propName, propValue]) => {
    if ((0, _withUniwindUtils.isColorClassProperty)(propName)) {
      const colorProp = (0, _withUniwindUtils.classToColor)(propName);
      if (props[colorProp] !== void 0) {
        return acc;
      }
      const className = propValue;
      const color = (0, _web.getWebStyles)(className, props, uniwindContext).accentColor;
      if (__DEV__ && !warnedOnce && (0, _utils.isDefined)(className) && className.trim() !== "" && color === void 0) {
        warnedOnce = true;
        _logger.Logger.warn(`className '${className}' was provided to extract accentColor but no color was found. Make sure the className includes a color utility (e.g., 'accent-red-500', 'accent-blue-600'). See https://docs.uniwind.dev/class-names#the-accent-prefix`);
      }
      acc.generatedProps[colorProp] = color !== void 0 ? (0, _web.formatColor)(color) : void 0;
      acc.classNames += `${className} `;
      return acc;
    }
    if ((0, _withUniwindUtils.isClassProperty)(propName)) {
      const styleProp = (0, _withUniwindUtils.classToStyle)(propName);
      acc.generatedProps[styleProp] ??= [];
      acc.generatedProps[styleProp][0] = {
        $$css: true,
        tailwind: propValue
      };
      return acc;
    }
    if ((0, _withUniwindUtils.isStyleProperty)(propName)) {
      acc.generatedProps[propName] ??= [];
      acc.generatedProps[propName][1] = propValue;
      return acc;
    }
    return acc;
  }, {
    generatedProps: {},
    classNames: ""
  });
  const [, rerender] = (0, _react.useReducer)(() => ({}), {});
  (0, _react.useLayoutEffect)(() => {
    const dispose = _web.CSSListener.subscribeToClassName(classNames, rerender);
    return dispose;
  }, [classNames]);
  return /* @__PURE__ */(0, _jsxRuntime.jsx)(Component, {
    ...props,
    ...generatedProps,
    dataSet: (0, _generateDataSet.generateDataSet)(props)
  });
};
const withManualUniwind = (Component, options) => props => {
  const uniwindContext = (0, _context.useUniwindContext)();
  const {
    generatedProps,
    classNames
  } = Object.entries(options).reduce((acc, [propName, option]) => {
    const className = props[option.fromClassName];
    if (className === void 0) {
      return acc;
    }
    if (option.styleProperty !== void 0) {
      if (props[propName] !== void 0) {
        return acc;
      }
      const value = (0, _web.getWebStyles)(className, props, uniwindContext)[option.styleProperty];
      const transformedValue = value !== void 0 && option.styleProperty.toLowerCase().includes("color") ? (0, _web.formatColor)(value) : value;
      acc.classNames += `${className} `;
      acc.generatedProps[propName] = transformedValue;
      return acc;
    }
    acc.generatedProps[propName] = [{
      $$css: true,
      tailwind: className
    }, props[propName]];
    return acc;
  }, {
    generatedProps: {},
    classNames: ""
  });
  const [, rerender] = (0, _react.useReducer)(() => ({}), {});
  (0, _react.useLayoutEffect)(() => {
    const dispose = _web.CSSListener.subscribeToClassName(classNames, rerender);
    return dispose;
  }, [classNames]);
  return /* @__PURE__ */(0, _jsxRuntime.jsx)(Component, {
    ...props,
    ...generatedProps,
    dataSet: (0, _generateDataSet.generateDataSet)(props)
  });
};
