// uniwind@1.9.0/dist/common/bundler/css-processor/addMetaToStylesTemplate.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addMetaToStylesTemplate = void 0;
var _consts = require("@/common/consts");
var _utils = require("@/common/utils");
var _serialize = require("./serialize");
var _utils2 = require("./utils");
const extractVarsFromString = value => {
  const varsIndexes = [...value.matchAll(/vars\[/g)].map(m => m.index);
  return varsIndexes.map(index => {
    const afterIndex = value.slice(index + 5);
    const closingIndex = afterIndex.indexOf("]");
    const varName = afterIndex.slice(0, closingIndex);
    return varName.replace(/[`"\\]/g, "");
  });
};
const makeSafeForSerialization = value => {
  if (value === null) {
    return null;
  }
  if (typeof value === "string") {
    return `"${value}"`;
  }
  return value;
};
const hasThemedVarDependency = (varName, Processor, visited = /* @__PURE__ */new Set()) => {
  if (visited.has(varName)) {
    return false;
  }
  visited.add(varName);
  const isScopedVar = Object.values(Processor.scopedVars).some(scopedVars => varName in scopedVars);
  if (isScopedVar) {
    return true;
  }
  const globalVarValue = Processor.vars[varName];
  if (typeof globalVarValue !== "string") {
    return false;
  }
  return extractVarsFromString(globalVarValue).some(usedVarName => {
    return hasThemedVarDependency(usedVarName, Processor, visited);
  });
};
const addMetaToStylesTemplate = (Processor, currentPlatform) => {
  const stylesheetsEntries = Object.entries(Processor.stylesheets).map(([className, stylesPerMediaQuery]) => {
    const styles = stylesPerMediaQuery.map((style, index) => {
      const {
        platform,
        rtl,
        theme,
        orientation,
        minWidth,
        maxWidth,
        colorScheme,
        important: _,
        importantProperties,
        active,
        focus,
        disabled,
        dataAttributes,
        ...rest
      } = style;
      const entries = Object.entries(rest).flatMap(([property, value]) => Processor.RN.cssToRN(property, value)).map(([property, value]) => [`"${property}"`, `function(vars) { return ${(0, _serialize.serialize)(value)} }`]);
      if (platform) {
        const isTV = currentPlatform === _consts.Platform.AndroidTV || currentPlatform === _consts.Platform.AppleTV;
        const commonPlatform = isTV ? _consts.Platform.TV : _consts.Platform.Native;
        if (platform !== commonPlatform && platform !== currentPlatform) {
          return null;
        }
      }
      if (entries.length === 0) {
        return null;
      }
      const dependencies = [];
      const stringifiedEntries = JSON.stringify(entries);
      const usedVars = extractVarsFromString(stringifiedEntries);
      const isUsingThemedVar = usedVars.some(usedVarName => hasThemedVarDependency(usedVarName, Processor));
      if (usedVars.length > 0) {
        dependencies.push(_consts.StyleDependency.Variables);
      }
      if (theme !== null || isUsingThemedVar || stringifiedEntries.includes("rt.lightDark")) {
        dependencies.push(_consts.StyleDependency.Theme);
      }
      if (orientation !== null) {
        dependencies.push(_consts.StyleDependency.Orientation);
      }
      if (rtl !== null) {
        dependencies.push(_consts.StyleDependency.Rtl);
      }
      if (Number(minWidth) !== 0 || Number(maxWidth) !== Number.MAX_VALUE || stringifiedEntries.includes("rt.screen")) {
        dependencies.push(_consts.StyleDependency.Dimensions);
      }
      if (stringifiedEntries.includes("rt.insets")) {
        dependencies.push(_consts.StyleDependency.Insets);
      }
      if (stringifiedEntries.includes("rt.fontScale")) {
        dependencies.push(_consts.StyleDependency.FontScale);
      }
      return {
        entries,
        minWidth,
        maxWidth,
        theme: makeSafeForSerialization(theme),
        orientation: makeSafeForSerialization(orientation),
        rtl,
        colorScheme: makeSafeForSerialization(colorScheme),
        native: platform !== null,
        dependencies: dependencies.length > 0 ? dependencies : null,
        index,
        className: makeSafeForSerialization(className),
        active,
        focus,
        disabled,
        importantProperties: importantProperties?.map(property => property.startsWith("--") ? property : (0, _utils2.toCamelCase)(property)).map(makeSafeForSerialization) ?? [],
        dataAttributes,
        complexity: [minWidth !== 0, theme !== null, orientation !== null, rtl !== null, platform !== null, active !== null, focus !== null, disabled !== null, dataAttributes !== null].filter(Boolean).length
      };
    });
    const filteredStyles = styles.filter(_utils.isDefined);
    if (filteredStyles.length === 0) {
      return null;
    }
    return [className, filteredStyles];
  }).filter(_utils.isDefined);
  const stylesheets = Object.fromEntries(stylesheetsEntries);
  return stylesheets;
};
exports.addMetaToStylesTemplate = addMetaToStylesTemplate;
