// uniwind@1.9.0/dist/common/bundler/css-processor/color.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Color = void 0;
var _culori = require("culori");
var _logger = require("../logger");
var _utils = require("./utils");
class Color {
  constructor(Processor) {
    this.Processor = Processor;
  }
  toRgb = (0, _culori.converter)("rgb");
  black = "#000000";
  logger = new _logger.Logger("Color");
  processColor(color) {
    if (typeof color === "string") {
      const parsed = (0, _culori.parse)(color);
      if (parsed === void 0) {
        const colorFunction = color.match(/^([a-zA-Z_$][a-zA-Z0-9_$]*)\s*\(/)?.[1];
        if (colorFunction === void 0) {
          this.logger.error(`Failed to convert color ${color}`);
          return this.black;
        }
        const colorValue = color.replace(colorFunction, "").slice(1, -1);
        return `rt.parseColor("${colorFunction}", ${colorValue})`;
      }
      return this.format(parsed);
    }
    try {
      if (color.type === "currentcolor") {
        return 'vars["currentColor"]?.(vars)';
      }
      if (color.type === "rgb" || color.type === "srgb") {
        const alpha = typeof color.alpha === "number" ? color.alpha : (0, _utils.pipe)(color.alpha)(x => this.Processor.CSS.processValue(x), Number, x => isNaN(x) ? 1 : x);
        return this.format({
          r: color.r / 255,
          g: color.g / 255,
          b: color.b / 255,
          alpha,
          mode: "rgb"
        });
      }
      const colorType = color.type === "display-p3" ? "p3" : color.type;
      const result = this.toRgb({
        mode: colorType,
        ...color
      });
      return this.format(result);
    } catch {
      this.logger.error(`Failed to convert color ${JSON.stringify(color)}`);
      return this.black;
    }
  }
  isColor(value) {
    return typeof value === "string" && (0, _culori.parse)(value) !== void 0;
  }
  format(color) {
    if (color.alpha === 1 || color.alpha === void 0) {
      return (0, _culori.formatHex)(color);
    }
    return (0, _culori.formatHex8)(color);
  }
}
exports.Color = Color;
