// uniwind@1.9.0/dist/common/bundler/css-processor/functions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Functions = void 0;
var _logger = require("../logger");
var _utils = require("./utils");
class Functions {
  constructor(Processor) {
    this.Processor = Processor;
  }
  logger = new _logger.Logger("Functions");
  processCalc(calc) {
    switch (calc.type) {
      case "sum":
        {
          const sum = calc.value.map(x => this.processCalc(x)).join(" + ");
          return this.tryEval(sum);
        }
      case "value":
        return this.Processor.CSS.processValue(calc.value);
      case "function":
        return this.Processor.CSS.processValue(calc.value);
      case "number":
        return String(calc.value);
      default:
        this.logger.warn(`Unsupported calc type - ${calc.type}`);
        return "";
    }
  }
  processFunction(fn) {
    if (typeof fn !== "object") {
      this.logger.warn(`Unsupported function - ${fn}`);
      return fn;
    }
    if (fn.name === "calc") {
      const calc = this.Processor.CSS.processValue(fn.arguments);
      return (0, _utils.pipe)(calc)(String, x => this.Processor.Functions.tryEval(x));
    }
    if (fn.name === "cubic-bezier") {
      const cubicArguments = (0, _utils.pipe)(this.Processor.CSS.processValue(fn.arguments))(String, x => x.replace(/,\s/g, ","));
      return `rt.cubicBezier(${cubicArguments})`;
    }
    if (fn.name === "max") {
      return `Math.max(${this.Processor.CSS.processValue(fn.arguments)})`;
    }
    if (fn.name === "linear-gradient") {
      return this.Processor.CSS.processValue(fn.arguments);
    }
    if (fn.name === "color-mix") {
      return this.processColorMix(fn);
    }
    if (["rgb", "oklab", "oklch", "hsl", "hwb", "lab", "lch", "srgb"].includes(fn.name)) {
      const color = `${fn.name}(${this.Processor.CSS.processValue(fn.arguments)})`;
      return this.Processor.Color.processColor(color);
    }
    if (["blur", "brightness", "contrast", "grayscale", "hue-rotate", "invert", "opacity", "saturate", "sepia", "conic-gradient", "radial-gradient"].includes(fn.name)) {
      return '""';
    }
    if (["skewX", "skewY"].includes(fn.name)) {
      return `"${fn.name}(${this.Processor.CSS.processValue(fn.arguments).replace(/"/g, "")})"`;
    }
    if (fn.name === "hairlineWidth") {
      return "rt.hairlineWidth";
    }
    if (fn.name === "pixelRatio") {
      if (fn.arguments.length === 0) {
        return "rt.pixelRatio(1)";
      }
      return `rt.pixelRatio(${this.Processor.CSS.processValue(fn.arguments)})`;
    }
    if (fn.name === "fontScale") {
      if (fn.arguments.length === 0) {
        return "rt.fontScale(1)";
      }
      return `rt.fontScale(${this.Processor.CSS.processValue(fn.arguments)})`;
    }
    if (fn.name === "drop-shadow") {
      return void 0;
    }
    this.logger.warn(`Unsupported function - ${fn.name}`);
    return fn.name;
  }
  processMathFunction(name, value) {
    if (!Array.isArray(value)) {
      return `Math.${name}(${this.processCalc(value)})`;
    }
    const values = value.map(x => this.processCalc(x)).join(" , ");
    return `Math.${name}(${values})`;
  }
  tryEval(value) {
    const units = Array.from(value.replace(/"/g, "").match(/(?<![A-Za-z-])(?:%|deg|rad|grad|turn)(?=\s|$)/g) ?? []);
    if (units.length === 0) {
      return value;
    }
    if (new Set(units).size !== 1) {
      this.logger.error(`Invalid calc, you can't mix multiple units`);
      return value;
    }
    if (units.includes("%") && value.includes("+")) {
      this.logger.error(`Invalid calc, you can't mix % with other units`);
      return value;
    }
    const unit = units.at(0) ?? "";
    try {
      const numericValue = value.replace(/"/g, "").replace(new RegExp(unit, "g"), "");
      return new Function(`return ${numericValue} + '${unit}'`)();
    } catch {
      this.logger.error(`Invalid calc ${value}`);
      return value;
    }
  }
  processColorMix(fn) {
    const tokens = fn.arguments.map(arg => (0, _utils.pipe)(arg)(x => this.Processor.CSS.processValue(x), String, x => x.trim())).filter(token => !["", ",", "in", "srgb", "rgb", "hsl", "hwb", "lab", "lch", "oklab", "oklch"].includes(token.replace(/"/g, "")));
    return `rt.colorMix(${tokens.join(", ")})`;
  }
}
exports.Functions = Functions;
