// uniwind@1.9.0/dist/common/bundler/css-processor/processor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProcessorBuilder = void 0;
var _consts = require("@/common/consts");
var _utils = require("@/common/utils");
var _lightningcss = require("lightningcss");
var _color = require("./color");
var _css = require("./css");
var _functions = require("./functions");
var _mq = require("./mq");
var _rn = require("./rn");
var _units = require("./units");
var _var = require("./var");
class ProcessorBuilder {
  constructor(bundlerConfig) {
    this.bundlerConfig = bundlerConfig;
    this.vars["--uniwind-em"] = this.bundlerConfig.polyfills?.rem ?? 16;
  }
  stylesheets = {};
  vars = {};
  scopedVars = {};
  CSS = new _css.CSS(this);
  RN = new _rn.RN(this);
  Var = new _var.Var(this);
  MQ = new _mq.MQ(this);
  Color = new _color.Color(this);
  Units = new _units.Units(this);
  Functions = new _functions.Functions(this);
  meta = {};
  declarationConfig = this.getDeclarationConfig();
  transform(css) {
    (0, _lightningcss.transform)({
      filename: "tailwind.css",
      code: Buffer.from(css),
      visitor: {
        StyleSheet: styleSheet => styleSheet.rules.forEach(rule => {
          this.declarationConfig = this.getDeclarationConfig();
          this.parseRuleRec(rule);
        })
      }
    });
  }
  getDeclarationConfig() {
    return {
      className: null,
      rtl: null,
      mediaQueries: [],
      root: false,
      theme: null,
      active: null,
      focus: null,
      disabled: null,
      dataAttributes: null
    };
  }
  addDeclaration(declaration, important = false) {
    const isVar = this.declarationConfig.root || this.declarationConfig.className === null;
    const mq = this.MQ.processMediaQueries(this.declarationConfig.mediaQueries);
    const style = (() => {
      if (!isVar) {
        return this.stylesheets[this.declarationConfig.className]?.at(-1);
      }
      if (mq.platform !== null) {
        const platformKey = `${_consts.UNIWIND_PLATFORM_VARIABLES}${mq.platform}`;
        this.scopedVars[platformKey] ??= {};
        return this.scopedVars[platformKey];
      }
      if (this.declarationConfig.theme === null) {
        return this.vars;
      }
      const themeKey = `${_consts.UNIWIND_THEME_VARIABLES}${this.declarationConfig.theme}`;
      this.scopedVars[themeKey] ??= {};
      return this.scopedVars[themeKey];
    })();
    if (!isVar) {
      Object.assign(style, mq);
      style.importantProperties ??= [];
      style.rtl = this.declarationConfig.rtl;
      style.theme = mq.colorScheme ?? this.declarationConfig.theme;
      style.active = this.declarationConfig.active;
      style.focus = this.declarationConfig.focus;
      style.disabled = this.declarationConfig.disabled;
      style.dataAttributes = this.declarationConfig.dataAttributes;
      this.meta.className = this.declarationConfig.className;
    }
    if (declaration.property === "unparsed") {
      style[declaration.value.propertyId.property] = this.CSS.processValue(declaration.value.value);
      if (!isVar && important) {
        style.importantProperties.push(declaration.value.propertyId.property);
      }
      return;
    }
    if (declaration.property === "custom") {
      style[declaration.value.name] = this.CSS.processValue(declaration.value.value);
      if (!isVar && important) {
        style.importantProperties.push(declaration.value.name);
      }
      return;
    }
    style[declaration.property] = this.CSS.processValue(declaration.value, declaration.property);
    if (!isVar && important) {
      style.importantProperties.push(declaration.property);
    }
  }
  parseRuleRec(rule) {
    if (this.declarationConfig.className !== null) {
      const lastStyle = this.stylesheets[this.declarationConfig.className]?.at(-1);
      if (lastStyle !== void 0 && Object.keys(lastStyle).length > 0) {
        this.stylesheets[this.declarationConfig.className]?.push({});
      }
    }
    if (rule.type === "style") {
      rule.value.selectors.forEach(selector => {
        const [maybeClassNameSelector] = selector;
        const newClassName = maybeClassNameSelector?.type === "class" ? maybeClassNameSelector.name : void 0;
        if (newClassName !== void 0) {
          this.declarationConfig.className = newClassName;
          this.stylesheets[newClassName] ??= [];
          this.stylesheets[newClassName].push({});
          rule.value.declarations?.declarations?.forEach(declaration => this.addDeclaration(declaration));
          rule.value.declarations?.importantDeclarations?.forEach(declaration => this.addDeclaration(declaration, true));
          rule.value.rules?.forEach(rule2 => this.parseRuleRec(rule2));
          return;
        }
        let rtl = null;
        let theme = null;
        let active = null;
        let focus = null;
        let disabled = null;
        let dataAttributes = null;
        selector.forEach(selector2 => {
          if (selector2.type === "pseudo-class" && selector2.kind === "where") {
            selector2.selectors.forEach(selector3 => {
              selector3.forEach(selector4 => {
                if (selector4.type === "class" && this.bundlerConfig.themes.includes(selector4.name)) {
                  theme = selector4.name;
                }
                if (selector4.type === "pseudo-class" && selector4.kind === "dir") {
                  rtl = selector4.direction === "rtl";
                }
              });
            });
          }
          if (selector2.type === "pseudo-class" && selector2.kind === "active") {
            active = true;
          }
          if (selector2.type === "pseudo-class" && selector2.kind === "focus") {
            focus = true;
          }
          if (selector2.type === "pseudo-class" && selector2.kind === "disabled") {
            disabled = true;
          }
          if (selector2.type === "attribute" && selector2.operation === null && selector2.name.startsWith("data-")) {
            dataAttributes ??= {};
            dataAttributes[selector2.name] = `"true"`;
          }
          if (selector2.type === "attribute" && selector2.operation?.operator === "equal" && selector2.name.startsWith("data-")) {
            dataAttributes ??= {};
            dataAttributes[selector2.name] = `"${selector2.operation.value}"`;
          }
        });
        if ([rtl, theme, active, focus, disabled, dataAttributes].some(_utils.isDefined)) {
          this.declarationConfig.rtl ??= rtl;
          this.declarationConfig.theme ??= theme;
          this.declarationConfig.active ??= active;
          this.declarationConfig.focus ??= focus;
          this.declarationConfig.disabled ??= disabled;
          this.declarationConfig.dataAttributes ??= dataAttributes;
          rule.value.declarations?.declarations?.forEach(declaration => this.addDeclaration(declaration));
          rule.value.declarations?.importantDeclarations?.forEach(declaration => this.addDeclaration(declaration, true));
          rule.value.rules?.forEach(rule2 => this.parseRuleRec(rule2));
          this.declarationConfig.rtl = null;
          this.declarationConfig.theme = null;
          this.declarationConfig.active = null;
          this.declarationConfig.focus = null;
          this.declarationConfig.disabled = null;
          this.declarationConfig.dataAttributes = null;
          return;
        }
        selector.forEach(selectorToken => {
          if (selectorToken.type === "pseudo-class" && selectorToken.kind === "root") {
            this.declarationConfig.root = true;
            rule.value.declarations?.declarations?.forEach(declaration => this.addDeclaration(declaration));
            rule.value.declarations?.importantDeclarations?.forEach(declaration => this.addDeclaration(declaration, true));
            rule.value.rules?.forEach(rule2 => this.parseRuleRec(rule2));
          }
        });
      });
      return;
    }
    if (rule.type === "supports" && rule.value.condition.value !== "div > div") {
      rule.value.rules.forEach(rule2 => this.parseRuleRec(rule2));
      return;
    }
    if (rule.type === "media") {
      const {
        mediaQueries
      } = rule.value.query;
      this.declarationConfig.mediaQueries.push(...mediaQueries);
      rule.value.rules.forEach(rule2 => {
        this.parseRuleRec(rule2);
        this.declarationConfig = this.getDeclarationConfig();
      });
      return;
    }
    if (rule.type === "layer-block") {
      rule.value.rules.forEach(rule2 => this.parseRuleRec(rule2));
      return;
    }
    if (rule.type === "nested-declarations") {
      rule.value.declarations.declarations?.forEach(declaration => this.addDeclaration(declaration));
      rule.value.declarations.importantDeclarations?.forEach(declaration => this.addDeclaration(declaration, true));
      return;
    }
    if (rule.type === "property" && rule.value.initialValue) {
      this.vars[rule.value.name] = this.CSS.processValue(rule.value.initialValue);
    }
  }
}
exports.ProcessorBuilder = ProcessorBuilder;
